extras/gizmo/gizmo.js

```diff
@@ -0,0 +1,325 @@
+import {
+    math,
+    PROJECTION_PERSPECTIVE,
+    SORTMODE_NONE,
+    EventHandler,
+    Layer,
+    Entity,
+    Mat4,
+    Vec3
+} from 'playcanvas';
+
+// temporary variables
+const tmpV1 = new Vec3();
+const tmpM1 = new Mat4();
+
+const xstart = new Vec3();
+const xdir = new Vec3();
+
+// constants
+const GIZMO_LAYER_ID = 1e5;
+const MIN_GIZMO_SCALE = 1e-4;
+const PERS_SCALE_RATIO = 0.3;
+const ORTHO_SCALE_RATIO = 0.32;
+
+/**
+ * The base class for all gizmos.
+ */
+class Gizmo extends EventHandler {
+    /**
+     * Internal version of the gizmo size.
+     *
+     * @type {number}
+     * @private
+     */
+    _size = 1;
+
+    /**
+     * Internal version of the gizmo scale.
+     *
+     * @type {number}
+     * @protected
+     */
+    _scale = 1;
+
+    /**
+     * Internal version of coordinate space.
+     *
+     * @type {string}
+     * @protected
+     */
+    _coordSpace = 'world';
+
+    /**
+     * The application instance containing the gizmo.
+     *
+     * @type {import('playcanvas').AppBase}
+     */
+    app;
+
+    /**
+     * The camera entity that displays the gizmo.
+     *
+     * @type {Entity}
+     */
+    camera;
+
+    /**
+     * The graph nodes attached to the gizmo.
+     *
+     * @type {import('playcanvas').GraphNode}
+     */
+    nodes = [];
+
+    /**
+     * The root gizmo entity.
+     *
+     * @type {import('playcanvas').Entity}
+     */
+    gizmo;
+
+    /**
+     * @typedef IntersectData
+     * @property {import('./mesh-tri-data.js').MeshTriData[]} meshTriDataList -
+     * The array of {@link MeshTriData}
+     * @property {import('playcanvas').GraphNode} parent - The mesh parent node.
+     * @property {import('playcanvas').MeshInstance[]} meshInstances -
+     * array of mesh instances for rendering
+     */
+    /**
+     * The intersection data object.
+     *
+     * @type {IntersectData[]}
+     */
+    intersectData = [];
+
+    /**
+     * Creates a new Gizmo object.
+     *
+     * @param {import('playcanvas').AppBase} app - The application instance.
+     * @param {import('playcanvas').CameraComponent} camera - The camera component.
+     * @example
+     * const gizmo = new pcx.Gizmo(app, camera);
+     */
+    constructor(app, camera) {
```

Comment:
do we need a `destroy` function too? That would remove the event handles from the window, remove the gizmo entity from the hierarchy root and similar.

Reply:
disconnect from app.on('update' ...

Reply:
we actually need to make sure all mesh instances get destroyed too, otherwise we leak GPU memory that is not garbage collected